### PR TITLE
Feature/989 get entity attribute

### PR DESCRIPTION
**Pull request by crbrox**
@fgalan 
@jmcanterafonseca

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -1,26 +1,41 @@
 FORMAT: 1A
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: This is weird... this PR doesn't include any modification to .apib file. Thus, why this is shown in the PR?

File: src/lib/apiTypesV2/Attribute.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+#include "common/tag.h"
+#include "apiTypesV2/Attribute.h"
+#include "ngsi10/QueryContextResponse.h"
+
+
+/* ****************************************************************************
+*
+* Attribute::render -
+*/
+std::string Attribute::render(ConnectionInfo* ciP, RequestType requestType, bool comma)
```
Comment (fgalan) [inline src/lib/apiTypesV2/Attribute.cpp]: v2 render() should be as "lean" as possible. Thus, I think that ciP and requestType should be removed (actually, I think they are not used in the function body). In addition, maybe the name of this function should be `toJson()` (not sure, but I think that APIv2 renders are using that name... check Ken's classes in apiTypesV2/).

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v1/updateContext, creating 5 entities
+# 02. GET /v2/entities/E1/attrs/T1
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: To be coherent with the entity creation, it should be:

GET /v2/entities/E1/attrs/A0

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+echo "02. GET /v2/entities/E1/attrs/NONE"
+echo "=================================="
+orionCurl --url '/v2/entities/E1/attrs/NONE' --json
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: It should be "/v2/entities/E1/attrs/A0".

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      ]
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+echo "02. GET /v2/entities/E1/attrs/NONE"
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: It should be:

GET /v2/entities/E1/attrs/A0

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+payload='{
+  "contextElements": [
+    {
+      "id":   "E1",
+      "type": "T1",
+      "attributes": [
+          {
+              "name": "A0",
+              "type": "number",
+              "value": ""
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Don't use an empty value... Note that error in the updateContextResponse you get if you do so.

File: src/lib/apiTypesV2/Attribute.cpp (hunk 2 above)
Comment (crbrox) [inline src/lib/apiTypesV2/Attribute.cpp]: I think it's the same https://github.com/telefonicaid/fiware-orion/blame/develop/src/lib/apiTypesV2/Entity.cpp#L58

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 3 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed c1a637660879500

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 4 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed c1a637660879500

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 5 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed c1a637660879500

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 6 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed c1a637660879500

File: src/lib/apiTypesV2/Attribute.cpp (hunk 2 above)
Comment (fgalan) [inline src/lib/apiTypesV2/Attribute.cpp]: NTC

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+echo "02. GET /v2/entities/E1/attrs/A0"
+echo "=================================="
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Line with '=' should be adjusted to the length of the step description line (i.e. delete 2 final '=').

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+GET Entity attribute -- Too many entities
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v1/updateContext, creating 5 entities
```
Comment (fgalan) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: ...2 entities

File: CHANGES_NEXT_RELEASE [hunk 9]
```diff
@@ -8,4 +8,5 @@ Fix: remove unconditional tracing on xmlTreat()/jsonTreat() which was impacting
 Fix: use NOSIGNAL as libcurl option to avoid crashes in that library. (Issue #1016)
 Add: CLI paramter -cprForwardLimit to cap the number of forwarding request for a single client request. (Issue #1016).
 Add: New operation "GET /v2/entities/[id]" (Issue #950)
+Add: New operation "GET /v2/entities/[id]/attrs/[attr]" (Issue #989)
```
Comment (jmcanterafonseca) [inline CHANGES_NEXT_RELEASE]: nit: GET /v2/entities/{entityId}/attrs/{attrName} better to use URL Templates https://tools.ietf.org/html/rfc6570

File: src/lib/serviceRoutinesV2/getEntityAttribute.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
```
Comment (jmcanterafonseca) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: nit: Ken :)

File: src/lib/serviceRoutinesV2/getEntityAttribute.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+
+  // 02. Call standard op postQueryContext
+  postQueryContext(ciP, components, compV, parseDataP);
+
+
+  // 03. Render entity attribute response
+  attribute.fill(&parseDataP->qcrs.res, compV[4]);
+  answer = attribute.render(ciP, EntityAttributeResponse);
+  if (attribute.errorCode.error == "TooManyResults")
```
Comment (jmcanterafonseca) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: Error names should be declared as constants, shouldn't they?

Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: Another alternative is to implement a "translation function" which takes as input the code and returns the string with the error, similar to `httpStatusCodeString(HttpStatusCode code)`.

Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: I think to improve in a further PR. 

AFAICS, currently we are using strings
https://github.com/telefonicaid/fiware-orion/blob/develop/src/lib/apiTypesV2/Entities.cpp#L39
https://github.com/telefonicaid/fiware-orion/blob/develop/src/lib/apiTypesV2/Entities.cpp#L131

Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: Let's move it to a tech debt issue.

Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: This is the tech debt issue: https://github.com/telefonicaid/fiware-orion/issues/1047

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 7 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed a5d388e

File: test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test (hunk 8 above)
Comment (crbrox) [inline test/functionalTest/cases/0989_GET_v2_entity_attribute/GET_v2_entity_attribute_too_many.test]: Fixed a5d388e

File: src/lib/serviceRoutinesV2/getEntityAttribute.cpp (hunk 10 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: Fixed a5d388e

File: CHANGES_NEXT_RELEASE (hunk 9 above)
Comment (crbrox) [inline CHANGES_NEXT_RELEASE]: Fixed 9894f06665

**Comment by fgalan** [comment]
LGTM (passing the ball to @jmcanterafonseca for final LGTM)

File: src/lib/serviceRoutinesV2/getEntityAttribute.cpp (hunk 11 above)
Comment (jmcanterafonseca) [inline src/lib/serviceRoutinesV2/getEntityAttribute.cpp]: ok on my side

NTC in this PR

**Comment by jmcanterafonseca** [comment]
LGTM

**Pull request merged**